fv/wireguard_test.go

```diff
@@ -0,0 +1,455 @@
+	var (
+		infra   infrastructure.DatastoreInfra
+		felixes []*infrastructure.Felix
+	)
+
+	BeforeEach(func() {
+		// Setup a single node cluster.
+		const nodeCount = 1
+
+		infra = getInfra()
+		felixes, _ = infrastructure.StartNNodeTopology(nodeCount, wireguardTopologyOptions(), infra)
+
+		// Install a default profile that allows all ingress and egress, in the absence of any Policy.
+		infra.AddDefaultAllow()
+
+		felixes[0].TriggerDelayedStart()
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			felixes[0].Exec("ip", "link")
+			felixes[0].Exec("wg")
+		}
+
+		felixes[0].Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+		infra.Stop()
+	})
+
+	It("no Wireguard device exists", func() {
+		Eventually(func() string {
+			out, _ := felixes[0].ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+			return out
+		}, "10s", "100ms").Should(BeEmpty())
+	})
+
+	It("no wg tool info exists", func() {
+		Eventually(func() string {
+			out, _ := felixes[0].ExecOutput("wg")
+			return out
+		}, "10s", "100ms").Should(BeEmpty())
+	})
+})
+
+func wireguardTopologyOptions() infrastructure.TopologyOptions {
+	topologyOptions := infrastructure.DefaultTopologyOptions()
+
+	// For attaching tcpdump and waiting on calico-node to be ready.
+	topologyOptions.DelayFelixStart = true
+	// WireGuard doesn't support IPV6, disable it.
+	topologyOptions.EnableIPv6 = false
+	// Assigning workload IPs using IPAM api.
+	topologyOptions.IPIPRoutesEnabled = false
+	// Enable WireGuard.
+	felixConfig := api.NewFelixConfiguration()
+	felixConfig.SetName("default")
+	enabled := true
+	felixConfig.Spec.WireguardEnabled = &enabled
+	topologyOptions.InitialFelixConfiguration = felixConfig
+
+	// Debugging.
+	//topologyOptions.ExtraEnvVars["FELIX_DebugUseShortPollIntervals"] = "true"
+	//topologyOptions.FelixLogSeverity = "debug"
+
+	return topologyOptions
+}
+
+func enableDisableWireguard(client clientv3.Interface, enabled bool) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	felixConfig, err := client.FelixConfigurations().Get(ctx, "default", options.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	felixConfig.Spec.WireguardEnabled = &enabled
+	felixConfig, err = client.FelixConfigurations().Update(ctx, felixConfig, options.SetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func wireguardRoutingRuleRouteEntryExists(felix *infrastructure.Felix, exp *regexp.Regexp, entry string) bool {
+	// Get WireGuard route-table index.
+	rule, err := felix.ExecOutput("ip", "rule", "show", "pref", wireguardRoutingRulePriorityDefault)
+	Expect(err).NotTo(HaveOccurred())
+	rule = strings.TrimSpace(rule)
+	if rule == "" {
+		return false
+	}
+	Expect(rule).To(MatchRegexp(`\d+:\s+from [\d\.]+/\d+ fwmark 0/0x\d+ lookup \d+`))
```

Comment:
This is probably ok for the first iteration, but I think this check is ultimately too restrictive.  Won't this break if there are multiple matching rules?

We can probably also validate the CIDR and fwmark in these rules are as expected. 

Fine for the first pass though.  We can refine the checks after these go in.

Reply:
For the FV setup, I don't see why there will be multiple rules, unless I am missing something.

>We can probably also validate the CIDR and fwmark in these rules are as expected.

Good point. I can look into extending the regex.

Reply:
@realgaurav - resolved?

Reply:
yup, it has been addressed in the latest update. I didn't mark any issue resolved because @robbrockbank asked me not to.

Reply:
okay. So back to you @robbrockbank - sounds like all issued are resolved, but we're not able to track whether issues are resolved, by design.